### PR TITLE
R-Py Parity for AI-ML / KNN Classifier Page

### PR DESCRIPTION
**Pull request by kvdesai**

### CODE REVIEW

File: r/2021-07-21-ml-knn.rmd [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+
+We will use [Tidymodels](https://www.tidymodels.org/) for training our model and for loading and splitting data. Tidymodels is a popular Machine Learning (ML) library that offers various tools for creating and training ML algorithms, feature engineering, data cleaning, and evaluating and testing models.
+
+We will train a [k-Nearest Neighbors (kNN)](https://parsnip.tidymodels.org/reference/nearest_neighbor.html) classifier. First, the model records the label of each training sample. Then, whenever we give it a new sample, it will look at the `k` closest samples from the training set to find the most common label, and assign it to our new sample.
+
+
+### Display training and test splits
+
+We get the synthetic data that form the shape of a moon. We then split it into a training and testing set. data that form the shape of a moon. We then split it into a training and testing set. Finally, we display the ground truth labels using [a scatter plot](https://plotly.com/r/line-and-scatter/).
```
Comment (HammadTheOne) [inline r/2021-07-21-ml-knn.rmd:17]:
```suggestion
Using Tidymodels, we first generate synthetic data that form the shape of a moon. We then split it into a training and testing set. Finally, we display the ground truth labels using [a scatter plot](https://plotly.com/r/line-and-scatter/).
```

File: r/2021-07-21-ml-knn.rmd [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+                             line = list(color = 'black',width = 1)))
+
+fig
+```
+
+### Visualize predictions on test split
+
+Now, we train the kNN model on the same training data displayed in the previous graph. Then, we predict the confidence score of the model for each of the data points in the test set. We will use shapes to denote the true labels, and the color will indicate the confidence of the model for assign that score.
+
+Notice that `scatter` only requires one function call to plot both negative and positive labels, and can additionally set a continuous color scale based on the `yscore` output by our kNN model.
```
Comment (HammadTheOne) [inline r/2021-07-21-ml-knn.rmd:52]:
```suggestion
Notice that `plot_ly` only requires one function call to plot both negative and positive labels, and can additionally set a continuous color scale based on the `yscore` output by our kNN model.
```

File: r/2021-07-21-ml-knn.rmd [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+We get the synthetic data that form the shape of a moon. We then split it into a training and testing set. data that form the shape of a moon. We then split it into a training and testing set. Finally, we display the ground truth labels using [a scatter plot](https://plotly.com/r/line-and-scatter/).
+
+In the graph, we display all the negative labels as squares, and positive labels as circles. We differentiate the training and test set by adding a dot to the center of test data.
+
+```{r}
+library(tidyverse)
+library(tidymodels)
+library(plotly)
+
+make_moons <- read.csv(file = "make_moons.csv")
```
Comment (HammadTheOne) [inline r/2021-07-21-ml-knn.rmd:26]:
```suggestion
make_moons <- read.csv(file = "data/make_moons.csv")
```

File: r/2021-07-21-ml-knn.rmd [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+  add_trace(data = make_moons, x = ~X1, y = ~X2, type = 'scatter', mode = 'markers', symbol = ~y ,
+            marker = list(size = 12,
+                          color = 'lightyellow',
+                          line = list(color = 'black',width = 1)))
+fig
+```
+
+## Multi-class prediction confidence with [`Heatmap`](https://plotly.com/r/heatmaps/)
+
+It is also possible to visualize the prediction confidence of the model using [heatmaps](https://plotly.com/r/heatmaps/). In this example, you can see how to compute how confident the model is about its prediction at every point in the 2D grid. Here, we define the confidence as the difference between the highest score and the score of the other classes summed, at a certain point.
```
Comment (HammadTheOne) [inline r/2021-07-21-ml-knn.rmd:209]:
```suggestion
It is also possible to visualize the prediction confidence of the model using [heatmaps](https://plotly.com/r/heatmaps/). In this example, you can see how to compute how confident the model is about its prediction at every point in the 2D grid. Here, we define the confidence as the difference between the highest score and the sum of the score of the other classes, at a certain point.
```

Minor suggestion - I know it's different from the Python docs, but it flows a bit better.

File: r/2021-07-21-ml-knn.rmd [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+
+```{r}
+library(pracma)
+library(plotly)
+library(tidyverse)
+library(tidymodels)
+library(plyr)
```
Comment (HammadTheOne) [inline r/2021-07-21-ml-knn.rmd:216]: There's a warning here about loading `plyr` after `dplyr`. Doesn't look like it's causing problems, but maybe we should load `plyr` as the first library.

**Review by HammadTheOne — APPROVED** [review]
This page looks great! I left some minor comments, one suggestion in particular is blocking because the dataset cannot be loaded. 

Other than that, looks good to me once the comments are addressed.

:woman_dancing:
